File: LexicalAnalysis.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace compiler
{
    public static class LexicalAnalysis
    {
        public static string keywords = " auto break case char const continue default do double else enum extern float for goto if int long register return short signed sizeof static struct switch typedef union unsigned void volatile while ";
        public static string digits = "0123456789";
        public static string letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_";
        public static string punctuations = "{}()[];,.";
        public static string signs = "+-/*=!<>&|";
        public static string strings = "\"";
        public static string delimiters = "\t\n ";
        public static string all = digits + letters + punctuations + signs + strings + delimiters;

        public static string state = "ST0";
        public static string stream = "";
        public static string identifiers = "";

        private static void PrintUngetch(char ch)
        {
            if (!delimiters.Contains(ch))
            {
                Console.WriteLine("ungetch");
            }
        }
        private static void PrintError()
        {
            Console.WriteLine("Error");
            stream = "";
        }
        private static void PrintNum(string type)
        {
            Console.WriteLine("{0}\t{1}", stream, type);
            stream = "";
        }
        private static void PrintStr()
        {
            Console.WriteLine("{0}\tSTR", stream);
            stream = "";
        }
        private static void PrintPunctuatuin(Char type)
        {
            Console.WriteLine("{0}\tPUNCTUATION", type);
            stream = "";
        }
        private static void PrintIdentifier()
        {
            Console.WriteLine("{0}\tIDENTIFIER", stream);
            if (identifiers.Contains(stream))
            {
                Console.WriteLine("false");
            }
            else
            {
                identifiers += stream + " ";
                Console.WriteLine("true");
            }
            stream = "";
        }
        private static void PrintOperator(string type)
        {
            Console.WriteLine("{0}\t{1}", stream, type);
            stream = "";
        }
        private static void PrintKeyword()
        {
            Console.WriteLine("{0}\tKEYWORD", stream);
            stream = "";
        }

        public static void SetState(char ch)
        {
            switch (state){
                case "ST0": ST0(ch); break;

                case "NUM1": NUM1(ch); break;
                case "NUM2": NUM2(ch); break;
                case "NUM3": NUM3(ch); break;
                case "NUM4": NUM4(ch); break;
                case "NUM5": NUM5(ch); break;
                case "NUM6": NUM6(ch); break;

                case "OPP": OPP(ch); break;
                case "OPS": OPS(ch); break;
                case "OPM": OPM(ch); break;
                case "OPA": OPA(ch); break;
                case "OPN": OPN(ch); break;
                case "OPL": OPL(ch); break;
                case "OPG": OPG(ch); break;
                case "OPZ": OPZ(ch); break;
                case "OPX": OPX(ch); break;
                    
                case "OPD1": OPD1(ch); break;
                case "OPD2": OPD2(ch); break;
                case "OPD3": OPD3(ch); break;
                case "OPD4": OPD4(ch); break;

                case "STR": STR(ch); break;
                case "ID": ID(ch); break;

                case "OTH": OTH(ch); break;
            }
        }

        private static void ST0(char ch)
        {
            if (digits.Contains(ch)) { state = "NUM1"; stream += ch; }
            else if (letters.Contains(ch)) { state = "ID"; stream += ch; }
            else if (punctuations.Contains(ch)) { PrintPunctuatuin(ch); }
            else if (delimiters.Contains(ch)) { }
            else if (ch == '"') { state = "STR"; stream += ch; }
            else if (ch == '+') { state = "OPP"; stream += ch; }
            else if (ch == '-') { state = "OPS"; stream += ch; }
            else if (ch == '*') { state = "OPM"; stream += ch; }
            else if (ch == '=') { state = "OPA"; stream += ch; }
            else if (ch == '!') { state = "OPN"; stream += ch; }
            else if (ch == '<') { state = "OPL"; stream += ch; }
            else if (ch == '>') { state = "OPG"; stream += ch; }
            else if (ch == '&') { state = "OPZ"; stream += ch; }
            else if (ch == '|') { state = "OPX"; stream += ch; }
            else if (ch == '/') { state = "OPD1"; stream += ch; }
            else { state = "OTH"; stream += ch; }
        }

        private static void NUM1(char ch)
        {
            if (digits.Contains(ch)) { stream += ch; }
            else if(ch == '.') { state = "NUM2"; stream += ch; }
            else if(ch == 'e' || ch == 'E') { state = "NUM4"; stream += ch; }
            else {
                PrintNum("INT");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void NUM2(char ch)
        {
            if (digits.Contains(ch)) { state = "NUM3"; stream += ch; }
            else
            {
                PrintError();
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void NUM3(char ch)
        {
            if (digits.Contains(ch)) { stream += ch; }
            else if (ch == 'e' || ch == 'E') { state = "NUM4"; stream += ch; }
            else
            {
                PrintNum("REAL");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void NUM4(char ch)
        {
            if (digits.Contains(ch)) { state = "NUM6"; stream += ch; }
            else if (ch == '+' || ch == '-') { state = "NUM5"; stream += ch; }
            else
            {
                PrintError();
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void NUM5(char ch)
        {
            if (digits.Contains(ch)) { state = "NUM6"; stream += ch; }
            else
            {
                PrintError();
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void NUM6(char ch)
        {
            if (digits.Contains(ch)) { stream += ch; }
            else
            {
                PrintNum("SCI");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }

        private static void OPP(char ch)
        {
            if (ch == '+') {
                stream += ch;
                PrintOperator("INC");
                state = "ST0";
            }
            else if (ch == '=')
            {
                stream += ch;
                PrintOperator("ADD_ASSIGN");
                state = "ST0";
            }
            else
            {
                PrintOperator("ADD");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPS(char ch)
        {
            if (ch == '-')
            {
                stream += ch;
                PrintOperator("DEC");
                state = "ST0";
            }
            else if (ch == '=')
            {
                stream += ch;
                PrintOperator("SUB_ASSIGN");
                state = "ST0";
            }
            else
            {
                PrintOperator("SUB");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPM(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("MUL_ASSIGN");
                state = "ST0";
            }
            else
            {
                PrintOperator("MUL");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPA(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("EQ");
                state = "ST0";
            }
            else
            {
                PrintOperator("ASSIGN");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPN(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("NE");
                state = "ST0";
            }
            else
            {
                PrintOperator("NOT");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPG(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("GE");
                state = "ST0";
            }
            else
            {
                PrintOperator("GT");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPL(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("LE");
                state = "ST0";
            }
            else
            {
                PrintOperator("LT");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPZ(char ch)
        {
            if (ch == '&')
            {
                stream += ch;
                PrintOperator("AND");
                state = "ST0";
            }
            else
            {
                PrintError();
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPX(char ch)
        {
            if (ch == '|')
            {
                stream += ch;
                PrintOperator("OR");
                state = "ST0";
            }
            else
            {
                PrintError();
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }

        private static void OPD1(char ch)
        {
            if (ch == '=')
            {
                stream += ch;
                PrintOperator("DIV_ASSIGN");
                state = "ST0";
            }
            else if (ch == '*')
            {
                stream += ch;
                state = "OPD2";
            }
            else if (ch == '/')
            {
                stream += ch;
                state = "OPD4";
            }
            else
            {
                PrintOperator("DIV");
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }
        private static void OPD2(char ch)
        {
            if (ch == '*')
            {
                stream += ch;
                state = "OPD3";
            }
            else
            {
                stream += ch;
            }
        }
        private static void OPD3(char ch)
        {
            if (ch == '/')
            {
                stream = "";
                state = "ST0";
            }
            else if (ch == '*')
            {
                stream += ch;
            }
            else
            {
                stream += ch;
                state = "OPD2";
            }
        }
        private static void OPD4(char ch)
        {
            if (ch == '\n')
            {
                stream = "";
                state = "ST0";
            }
            else
            {
                stream += ch;
            }
        }

        private static void STR(char ch)
        {
            if (ch == '\"')
            {
                stream += ch;
                PrintStr();
                state = "ST0";
            }
            else
            {
                stream += ch;
            }
        }
        private static void ID(char ch)
        {
            if (letters.Contains(ch) || digits.Contains(ch))
            {
                stream += ch;
            }
            else
            {
                if (keywords.Contains(" "+ stream + " "))
                {
                    PrintKeyword();
                }
                else
                {
                    PrintIdentifier();
                }
                PrintUngetch(ch);
                state = "ST0";
                SetState(ch);
            }
        }

        private static void OTH(char ch)
        {
            if (all.Contains(ch))
            {
                PrintError();
                state = "ST0";
                SetState(ch);
            }
            else
            {
                stream += ch;
            }
        }
    }
}

File: python/lexical_analysis.py
from simplesm.simplesm import SimpleSM


class CLexicalAnalysis(SimpleSM):
    keywords = ["auto", "break", "case", "char", "const", "continue", "default", "do", "double", "else", "enum",
                "extern", "float", "for", "goto", "if", "int", "long", "register", "return", "short", "signed",
                "sizeof", "static", "struct", "switch", "typedef", "union", "unsigned", "void", "volatile", "while"]
    digits = "0123456789"
    letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_"
    punctuations = "{}()[];,."
    signs = "+-/*=!<>&|"
    strings = "\""
    delimiters = "\t\n "
    all = digits + letters + punctuations + signs + strings + delimiters

    stream = ""
    identifiers = []

    start_state = "ST0"
    states = {
        "ST0": {
            "transitions": {
                "is_digit": {"state": "NUM1", "event": "update_stream"},
                "is_letter": {"state": "ID", "event": "update_stream"},
                "is_punctuation": {"state": "ST0", "event": "print_punctuation"},
                "is_delimiter": {"state": "ST0"},
                "\"": {"state": "STR", "event": "update_stream"},
                "+": {"state": "OPP", "event": "update_stream"},
                "-": {"state": "OPS", "event": "update_stream"},
                "*": {"state": "OPM", "event": "update_stream"},
                "=": {"state": "OPA", "event": "update_stream"},
                "!": {"state": "OPN", "event": "update_stream"},
                "<": {"state": "OPL", "event": "update_stream"},
                ">": {"state": "OPG", "event": "update_stream"},
                "&": {"state": "OPZ", "event": "update_stream"},
                "|": {"state": "OPX", "event": "update_stream"},
                "/": {"state": "OPD1", "event": "update_stream"},
                "other": {"state": "OTH", "event": "update_stream"}}},
        "NUM1": {
            "transitions": {
                "is_digit": {"state": "NUM1", "event": "update_stream"},
                ".": {"state": "NUM2", "event": "update_stream"},
                "e": {"state": "NUM4", "event": "update_stream"},
                "E": {"state": "NUM4", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_num1"}}},
        "NUM2": {
            "transitions": {
                "is_digit": {"state": "NUM3", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_error_ungetch"}}},
        "NUM3": {
            "transitions": {
                "is_digit": {"state": "NUM3", "event": "update_stream"},
                "e": {"state": "NUM4", "event": "update_stream"},
                "E": {"state": "NUM4", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_num3"}}},
        "NUM4": {
            "transitions": {
                "is_digit": {"state": "NUM6", "event": "update_stream"},
                "-": {"state": "NUM5", "event": "update_stream"},
                "+": {"state": "NUM5", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_error_ungetch"}}},
        "NUM5": {
            "transitions": {
                "is_digit": {"state": "NUM6", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_error_ungetch"}}},
        "NUM6": {
            "transitions": {
                "is_digit": {"state": "NUM6", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_num6"}}},
        "OPP": {
            "transitions": {
                "+": {"state": "ST0", "event": "do_opp_inc"},
                "=": {"state": "ST0", "event": "do_opp_add_assign"},
                "other": {"state": "ST0", "event": "do_opp_add"}}},
        "OPS": {
            "transitions": {
                "-": {"state": "ST0", "event": "do_ops_dec"},
                "=": {"state": "ST0", "event": "do_ops_sub_assign"},
                "other": {"state": "ST0", "event": "do_ops_sub"}}},
        "OPM": {
            "transitions": {
                "=": {"state": "ST0", "event": "do_opm_mul_assign"},
                "other": {"state": "ST0", "event": "do_opm_mul"}}},
        "OPA": {
            "transitions": {
                "=": {"state": "ST0", "event": "do_opa_eq"},
                "other": {"state": "ST0", "event": "do_opa_assign"}}},
        "OPN": {
            "transitions": {
                "=": {"state": "ST0", "event": "do_opn_ne"},
                "other": {"state": "ST0", "event": "do_opn_not"}}},
        "OPG": {
            "transitions": {
                "=": {"state": "ST0", "event": "do_opg_ge"},
                "other": {"state": "ST0", "event": "do_opg_gt"}}},
        "OPL": {
            "transitions": {
                "=": {"state": "ST0", "event": "do_opl_le"},
                "other": {"state": "ST0", "event": "do_opl_lt"}}},
        "OPZ": {
            "transitions": {
                "&": {"state": "ST0", "event": "do_opz_and"},
                "other": {"state": "ST0", "event": "do_error_ungetch"}}},
        "OPX": {
            "transitions": {
                "|": {"state": "ST0", "event": "do_opx_or"},
                "other": {"state": "ST0", "event": "do_error_ungetch"}}},
        "OPD1": {
            "transitions": {
                "*": {"state": "OPD2", "event": "update_stream"},
                "/": {"state": "OPD4", "event": "update_stream"},
                "=": {"state": "ST0", "event": "do_opd1_div_assign"},
                "other": {"state": "ST0", "event": "do_opd1_div"}}},
        "OPD2": {
            "transitions": {
                "*": {"state": "OPD3", "event": "update_stream"},
                "other": {"state": "OPD2", "event": "update_stream"}}},
        "OPD3": {
            "transitions": {
                "*": {"state": "OPD3", "event": "update_stream"},
                "/": {"state": "ST0", "event": "clear_stream"},
                "other": {"state": "OPD2", "event": "update_stream"}}},
        "OPD4": {
            "transitions": {
                "\n": {"state": "ST0", "event": "clear_stream"},
                "other": {"state": "OPD4", "event": "update_stream"}}},
        "STR": {
            "transitions": {
                "\"": {"state": "ST0", "event": "do_str"},
                "other": {"state": "STR", "event": "update_stream"}}},
        "ID": {
            "transitions": {
                "is_letter": {"state": "ID", "event": "update_stream"},
                "is_digit": {"state": "ID", "event": "update_stream"},
                "other": {"state": "ST0", "event": "do_id"}}},
        "OTH": {
            "transitions": {
                "is_in_all": {"state": "ST0", "event": "print_error"},
                "other": {"state": "OTH", "event": "update_stream"}}},
    }

    # type check
    def is_digit(self):
        if self.test_action in self.digits:
            return True
        return False

    def is_letter(self):
        if self.test_action in self.letters:
            return True
        return False

    def is_punctuation(self):
        if self.test_action in self.punctuations:
            return True
        return False

    def is_delimiter(self):
        if self.test_action in self.delimiters:
            return True
        return False

    def is_in_all(self):
        if self.test_action in self.all:
            return True
        return False

    def other(self):
        return True

    # prints
    def print_ungetch(self):
        if not self.test_action in self.delimiters:
            print("ungetch")

    def print_error(self):
        print("Error")
        self.stream = ""

    def print_num(self, _type):
        print("{}\t{}".format(self.stream, _type))
        self.stream = ""

    def print_str(self):
        print("{}\tSTR".format(self.stream))
        self.stream = ""

    def print_punctuation(self):
        print("{}\tPUNCTUATION".format(self.test_action))
        self.stream = ""

    def print_identifier(self):
        print("{}\tIDENTIFIER".format(self.stream))
        if self.stream in self.identifiers:
            print('false')
        else:
            print('true')
            self.identifiers.append(self.stream)
        self.stream = ""

    def print_operator(self, _type):
        print("{}\t{}".format(self.stream, _type))
        self.stream = ""

    def print_keyword(self):
        print("{}\tKEYWORD".format(self.stream))
        self.stream = ""

    # states definitions
    def update_stream(self):
        self.stream += self.test_action

    def clear_stream(self):
        self.stream = ""

    def do_error_ungetch(self):
        self.print_error()
        self.print_ungetch()
        self.perform(self.test_action)

    def do_num1(self):
        self.print_num('INT')
        self.print_ungetch()
        self.perform(self.test_action)

    def do_num3(self):
        self.print_num('REAL')
        self.print_ungetch()
        self.perform(self.test_action)

    def do_num6(self):
        self.print_num('SCI')
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opp_inc(self):
        self.update_stream()
        self.print_operator("INC")

    def do_opp_add_assign(self):
        self.update_stream()
        self.print_operator("ADD_ASSIGN")

    def do_opp_add(self):
        self.print_operator("ADD")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_ops_dec(self):
        self.update_stream()
        self.print_operator("DEC")

    def do_ops_sub_assign(self):
        self.update_stream()
        self.print_operator("SUB_ASSIGN")

    def do_ops_sub(self):
        self.print_operator("SUB")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opm_mul_assign(self):
        self.update_stream()
        self.print_operator("MUL_ASSIGN")

    def do_opm_mul(self):
        self.print_operator("MUL")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opa_eq(self):
        self.update_stream()
        self.print_operator("EQ")

    def do_opa_assign(self):
        self.print_operator("ASSIGN")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opn_ne(self):
        self.update_stream()
        self.print_operator("NE")

    def do_opn_not(self):
        self.print_operator("NOT")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opg_ge(self):
        self.update_stream()
        self.print_operator("GE")

    def do_opg_gt(self):
        self.print_operator("GT")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opl_le(self):
        self.update_stream()
        self.print_operator("LE")

    def do_opl_lt(self):
        self.print_operator("LT")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_opz_and(self):
        self.update_stream()
        self.print_operator("AND")

    def do_opx_or(self):
        self.update_stream()
        self.print_operator("OR")

    def do_opd1_div_assign(self):
        self.update_stream()
        self.print_operator("DIV_ASSIGN")

    def do_opd1_div(self):
        self.print_operator("DIV")
        self.print_ungetch()
        self.perform(self.test_action)

    def do_str(self):
        self.update_stream()
        self.print_str()

    def do_id(self):
        if self.stream in self.keywords:
            self.print_keyword()
        else:
            self.print_identifier()
        self.print_ungetch()
        self.perform(self.test_action)
